File: chaos_test.py
from locust import HttpUser, task, between
import random

class AegisChaosEngine(HttpUser):
    # Simulate realistic user delays (100ms to 500ms between clicks)
    wait_time = between(0.1, 0.5)

    @task
    def transfer_funds(self):
        # Generate a random integer between 1 and 100 and zero-pad it to 3 digits dynamically
        user_id = f"user-{random.randint(1, 100):03d}"

        payload = {
            "userId": user_id,
            "amount": round(random.uniform(10.0, 500.0), 2),
            "destinationAccount": f"012345{random.randint(1000, 9999)}",
            "destinationBankCode": "058"
        }

        with self.client.post("/api/v1/transfers/transfer", json=payload, catch_response=True) as response:
            # We treat 504 as a success because the Saga pattern successfully caught it
            if response.status_code in [200, 201, 202, 504]:
                response.success()
            elif response.status_code == 500:
                response.failure(f"System Crash: {response.text}")
            else:
                response.failure(f"Unexpected Status: {response.status_code}")

File: chaos_test.js
import http from 'k6/http';
import { check, sleep } from 'k6';

// The attack profile
export const options = {
    stages: [
        { duration: '10s', target: 50 },  // Ramp up to 50 concurrent users
        { duration: '30s', target: 500 }, // Hold the spike at 500 users
        { duration: '10s', target: 0 },   // Graceful scale down
    ],
};

export default function () {
    const url = 'http://localhost:8080/api/v1/transfers/transfer';

    // Generate a random user from 1 to 100 and pad it with leading zeros (e.g., 001, 042, 100)
    const randomUserNum = Math.floor(Math.random() * 100) + 1;
    const paddedUserNum = String(randomUserNum).padStart(3, '0');

    const payload = JSON.stringify({
        userId: `user-${paddedUserNum}`,
        amount: parseFloat((Math.random() * (500 - 10) + 10).toFixed(2)),
        destinationAccount: `012345${Math.floor(1000 + Math.random() * 9000)}`,
        destinationBankCode: '058',
    });

    const params = {
        headers: {
            'Content-Type': 'application/json',
        },
    };

    const res = http.post(url, payload, params);

    // Validate the response. 504 Gateway Timeout is an expected architectural behavior.
    check(res, {
        'Success or Controlled Timeout (200/201/202/504)': (r) => [200, 201, 202, 504].includes(r.status),
    });

    // sleep(Math.random() * (0.5 - 0.1) + 0.1);
}